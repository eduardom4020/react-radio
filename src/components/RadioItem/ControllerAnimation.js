import { useState, useCallback, useEffect } from 'react';
import { Animated, Easing } from 'react-native';

const useAnimationController = props => {
    const { shouldExpand } = props;

    const [ linearAnimation, ] = useState(new Animated.Value(0));

    const callLinear = useCallback(
        (expanding) => {
            linearAnimation.setValue(expanding === true ? 0 : 1)
            Animated.timing(
                linearAnimation,
                {
                    toValue: expanding === true ? 1 : 0,
                    duration: 150,
                    easing: Easing.linear
                }
            ).start()
        }, 
        [ linearAnimation ]
    );

    useEffect(
        () => callLinear(shouldExpand),
        [ callLinear, shouldExpand ]
    );

    const linearGrow = useCallback(
        (from, to) => {
            return linearAnimation.interpolate({
                inputRange: [0, 1],
                outputRange: [from, to]
            });
        }, 
        [ linearAnimation ]
    );    

    return {
        linearGrow
    };
};

export default useAnimationController;