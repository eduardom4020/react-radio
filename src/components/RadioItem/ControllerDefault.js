import { useState, useCallback, useEffect } from 'react';
import { Animated } from 'react-native';

const useController = props => {
    const { onActive } = props;
    
    const [ isActive, setActive ] = useState(false);
    // const [ animation, setAnimation ] = useState(new Animated.Value());

    const toggle = useCallback(
        () => {
            setActive(!isActive);

            if(!isActive && onActive) {
                onActive();
            }
        }, 
        [ isActive, onActive, setActive ]
    );

    // useEffect(
    //     () => {
    //         if(isActive) {
    //             animation.setValue(0);
    //             Animated.spring(
    //                 animation,
    //                 {
    //                     toValue: 200
    //                 }
    //             ).start(); 
    //         }
    //     },
    //     [ animation, isActive ]
    // );

    return {
        isActive,
        toggle,
        // animation
    };
};

export default useController;