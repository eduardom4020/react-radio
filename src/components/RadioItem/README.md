# RadioItem
>The main interactive component of the application. It is a composition of ListItem and RadioDetail components. The final result is a item with toggle action that triggers an smooth toggle animation, revealing the image when click on radio.


This component uses two Controllers (default and animation), that are logic modules, written using hooks, responsible for complex behaviours of the component. State manipulation, side effects, data fetch and more are examples of these behaviours. The reason wich they are separated controllers is that they do way distinct things, and one of them needs to use React Native's Animated API, turning this AnimationController exclusive for Native, but the DefaultController can still being able to reuse in any other React applicaition.


### Props
These are the props accepted by the component, pay attention to the description, inputs and behaviour:

---

##### id *(string | number)*
*The id of the radio. It should be unique.*

**example**: 
```
<RadioItem id='1' />
```

---

##### title *(string)*
*The name of the radio*.

**example**: 
```
<RadioItem title='Doge FM' />
```

---

##### frequency *(string | number)*
*A string or a number representing the frequency number wich the radio works with.*

**example**: 
```
<RadioItem frequency='123,4' />
```

---

##### tag *(string)*
*A tag that feeds the props imageTag of RadioDetail component.*

**example**: 
```
<RadioItem tag='doge.png' />
```

---

##### selectedRadio *(object)*
*The current radio being listened in the entire application. It is used when switching between radios, in order the close radios that aren't the selected one.*

**example**: 
```
<RadioItem selectedRadio={{id: '1', title: 'Doge FM', ...}} />
```

---

##### onToggle *(function)*
*A callback function that is triggered when radio is about to active. In this application is being used for update radio name on footer.*

**example**: 
```
<RadioItem onToggle={(id, isActive) => console.log(id, isActive)} />
```

---

**ATTENTION**: Any other props passed to this component will work as props for an ListItem component.