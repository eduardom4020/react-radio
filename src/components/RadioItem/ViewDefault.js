import React from 'react';
import styled from 'styled-components';
import { Text } from 'react-native';
import { ListItem } from '~/src/components';

const RadioItemViewDefault = props => {
    const { title='Some Radio', frequency='111,11' } = props;
    
    return (
        <ListItem color='transparent'>
            <Text style={{color: '#a3acbe', fontSize: 18, textAlign: 'left'}}>{title}</Text>
            <Text style={{color: '#a3acbe', fontWeight: 'bold', fontSize: 18, textAlign: 'right'}}>{frequency}</Text>
        </ListItem>
    );
};

export default RadioItemViewDefault;