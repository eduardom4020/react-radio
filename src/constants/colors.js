export const BLACK_LIKE = '#20202b';
export const DARK_GREY = '#2e2e3a';
export const MEDIUM_GREY = '#a3acbe';
export const MEDIUM_ORANGE = '#eeae61';
export const METALIC_BLACK = '#191313';