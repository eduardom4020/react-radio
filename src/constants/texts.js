export const APP_TITLE = 'STATIONS';