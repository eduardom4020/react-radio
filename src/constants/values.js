export const DEFAULT_IMAGE_BT_SIZE = 192;
export const DEFAULT_ICON_BT_SIZE = 24;
export const DEFAULT_RADIO_PLAYING = 'CURRENTLY PLAYING'