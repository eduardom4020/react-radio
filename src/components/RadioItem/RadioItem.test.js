import React from 'react';
import { Text, View } from 'react-native';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { DARK_GREY } from '~/src/constants/colors';
import RadioItem from '~/src/components/RadioItem';

describe('RadioToolbar', () => {
    test('Render?', () => {
        const component = shallow(
            <RadioItem />
        );

        expect(toJson(
            component, {
                noKey: false,
                mode: 'deep'
            }
        )).toMatchSnapshot();
    });

    test('Render with custom color?', () => {
        const component = shallow(
            <RadioItem color={DARK_GREY} />
        );

        expect(toJson(
            component, {
                noKey: false,
                mode: 'deep'
            }
        )).toMatchSnapshot();
    });

    
});