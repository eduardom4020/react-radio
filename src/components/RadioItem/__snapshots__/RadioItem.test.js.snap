// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioToolbar Render with custom color? 1`] = `
<Fragment>
  <AnimatedComponent
    style={
      Object {
        "height": 0,
        "width": "100%",
      }
    }
  >
    <RadioDetailViewDefault
      onTouchImage={[Function]}
    />
  </AnimatedComponent>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
  >
    <Styled(View)
      color="#2e2e3a"
    >
      <Styled(Text)>
        Some Radio
      </Styled(Text)>
      <Styled(Text)
        bold={true}
        toRight={true}
      >
        111,11
      </Styled(Text)>
    </Styled(View)>
  </TouchableOpacity>
</Fragment>
`;

exports[`RadioToolbar Render? 1`] = `
<Fragment>
  <AnimatedComponent
    style={
      Object {
        "height": 0,
        "width": "100%",
      }
    }
  >
    <RadioDetailViewDefault
      onTouchImage={[Function]}
    />
  </AnimatedComponent>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
  >
    <Styled(View)
      color="transparent"
    >
      <Styled(Text)>
        Some Radio
      </Styled(Text)>
      <Styled(Text)
        bold={true}
        toRight={true}
      >
        111,11
      </Styled(Text)>
    </Styled(View)>
  </TouchableOpacity>
</Fragment>
`;
